src/observer.jl

```diff
@@ -1,72 +1,114 @@
 """
-TomographyObserver is an implementation of an
+CircuitObserver is an implementation of an
 observer object (<:AbstractObserver) which
-implements custom measurements.
+implements custom measurements to perform
+at each layer of the circuit evolution.
 """
-struct TomographyObserver <: AbstractObserver
-  fidelity::Vector{Float64}
-  train_loss::Vector{Float64}
-  test_loss::Vector{Float64}
-  trace_preserving_distance::Vector{Float64}
-  frobenius_distance::Vector{Float64}
-  fidelity_bound::Vector{Float64}
-  function TomographyObserver()
-    return new([],[],[],[],[],[])
-  end
+struct Observer <: AbstractObserver
```

Comment:
Not trying to throw a wrench in things, I think the design now is good. However it kind of bothers me that we are storing two dictionaries, which mostly have the same keys.

I was thinking that we could just store one dictionary, where the entries are a pair of the function/measurement and the result, i.e.
```
struct Observer <: AbstractObserver
  measurements::Dict{String, Pair{Union{Nothing, Function, String, Tuple}, Any}
end
```
Then, "default" measurements that are added internally by a function (for example `tomography`) would be added as `obs.measurements["fidelity"] = nothing => []`. Let me know what you think, it shouldn't be too hard of a change to make (some small adjustments to the `Observer` constructor, `push!` function, and `measure!` function).

We should also provide helper functions, maybe something like `emptyresults!`, that empties the vectors of results so that you can run a new calculation with the same set of measurements. Not totally necessary right now, just putting it out there.

Also, what kind of measurement is of type `Pair{<:Function, <:Any}}}`?